mxcubeweb/core/components/harvester.py

```diff
@@ -0,0 +1,494 @@
+# -*- coding: utf-8 -*-
+import logging
+import gevent
+from typing import Optional
+
+from mxcubecore import queue_entry
+from mxcubecore import HardwareRepository as HWR
+
+from mxcubeweb.core.components.component_base import ComponentBase
+from mxcubeweb.core.components.queue import COLLECTED, UNCOLLECTED
+
+from mxcubecore.HardwareObjects.abstract.sample_changer import Crims
+
+
+# TO CONSIDER:
+# This should maybe be made into a adapter instead of a component
+class Harvester(ComponentBase):
+    def __init__(self, app, config):
+        super().__init__(app, config)
+
+    def init_signals(self):
+        from mxcubeweb.routes import signals
+
+        """Initialize hwobj signals."""
+        if HWR.beamline.harvester:
+            HWR.beamline.harvester.connect(
+                "stateChanged", signals.harvester_state_changed
+            )
+
+            HWR.beamline.harvester.connect(
+                "harvester_contents_update", signals.harvester_contents_update
+            )
+
+    def get_initial_state(self):
+        if HWR.beamline.harvester_maintenance is not None:
+            (
+                global_state,
+                cmdstate,
+                msg,
+            ) = HWR.beamline.harvester_maintenance.get_global_state()
+
+            cmds = HWR.beamline.harvester_maintenance.get_cmd_info()
+        else:
+            global_state = {}
+            cmdstate = "harvester maintenance controller not defined"
+            cmds = []
+            msg = ""
+
+        contents = self.get_harvester_contents()
+
+        try:
+            state = HWR.beamline.harvester.get_status().upper()
+        except Exception:
+            state = "OFFLINE"
+
+        initial_state = {
+            "state": state,
+            "contents": contents,
+            "global_state": {"global_state": global_state, "commands_state": cmdstate},
+            "cmds": {"cmds": cmds},
+            "msg": msg,
+            "plate_mode": HWR.beamline.diffractometer.in_plate_mode(),
+        }
+
+        return initial_state
+
+    def mount_from_harvester(self):
+        sc = HWR.beamline.sample_changer
+
+        try:
+            return sc.mount_from_harvester()
+        except AttributeError:
+            return False
+
+    def get_harvester_contents(self):
+        """Get the Harvester contents info
+
+        Return (Dict): Dict content, object name etc..
+        crystal_list and number of available pins"
+        """
+        if HWR.beamline.harvester:
+            root_name = HWR.beamline.harvester.__TYPE__
+            crystal_list = self.get_crystal_list()
+            room_temperature_mode = HWR.beamline.harvester.get_room_temperature_mode()
+            number_of_pins = HWR.beamline.harvester.get_number_of_available_pin()
+            contents = {
```

Comment:
It would be nice to model contents, so that we know exactly what it returns. Like we have done in a few places, for instance here, https://github.com/mxcube/mxcubeweb/blob/develop/mxcubeweb/core/models/adaptermodels.py